### PR TITLE
WIP format literal arg inlining

### PR DESCRIPTION
**Pull request by nyurik**
A rough draft of #10739 implementation. The goal is to allow any kind of literal string arguments to be inlined automatically.

```rust
format!("hello {}", "world")
// is replaced with
format!("hello world")
```

```
changelog: [`uninlined_format_args`]: support for inlining string literals
```

### CODE REVIEW

File: clippy_lints/src/format_args.rs [hunk 1]
```diff
@@ -375,6 +381,21 @@ fn check_one_arg(
         fixes.push((pos_span, replacement));
         fixes.push((arg_span, String::new()));
         true  // successful inlining, continue checking
+    } else if !matches!(arg.kind, FormatArgumentKind::Captured(_))
+        && let Some(FormatPlaceholder{span: Some(placeholder_span), ..}) = placeholder
+        && let rustc_ast::ExprKind::Lit(lit) = &arg.expr.kind
+        && lit.kind == LitKind::Str // FIXME: lit.kind must match the format string kind
```
Comment (llogiq) [inline clippy_lints/src/format_args.rs:387]: This isn't clear to me. What other kinds would there be?

File: tests/ui/uninlined_format_args.fixed [hunk 2]
```diff
@@ -262,3 +262,17 @@ fn tester2() {
         local_i32,
     };
 }
+
+fn literals() {
```
Comment (llogiq) [inline tests/ui/uninlined_format_args.fixed:266]: I worry about cases where the format string and literal value come from different macro expansions. We should have a test where that is the case.

File: tests/ui/uninlined_format_args.stderr [hunk 3]
```diff
@@ -177,6 +177,18 @@ LL -     println!("{:.1}", local_f64);
...
+   |
+LL |     println!("Hello {} is {:.*}", "x", local_i32, local_f64);
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+help: change this to
+   |
+LL -     println!("Hello {} is {:.*}", "x", local_i32, local_f64);
+LL +     println!("Hello x is {local_f64:.local_i32$}");
+   |
+
```
Comment (llogiq) [inline tests/ui/uninlined_format_args.stderr:191]: Looks pretty good. I have just one request for clarification and I'd like to see a test with macros.

**Review by llogiq — COMMENTED** [review]
Looks good to me. I have but one request for clarification and I'd like to see a test with macros.

File: clippy_lints/src/format_args.rs (hunk 1 above)
Comment (nyurik) [reply clippy_lints/src/format_args.rs:387]: there are also [raw strings](https://doc.rust-lang.org/reference/tokens.html#raw-string-literals)

Comment (nyurik) [reply clippy_lints/src/format_args.rs:387]: And possibly byte strings too?  Not certain

File: tests/ui/uninlined_format_args.fixed (hunk 2 above)
Comment (nyurik) [reply tests/ui/uninlined_format_args.fixed:266]: Yeah, i will need tons more of tests for this - some already exist in
* https://github.com/rust-lang/rust-clippy/blob/master/tests/ui/write_literal.rs
* https://github.com/rust-lang/rust-clippy/blob/master/tests/ui/write_literal_2.rs
* https://github.com/rust-lang/rust-clippy/blob/master/tests/ui/write_with_newline.rs

Could you give an example of what specific macro expansion where you thinking about?

**Comment by Alexendoo** [comment]
I'd say this should be under a new lint name, although it's similar what it's doing is different enough to the existing `uninlined_format_args`

The existing implementation can also be used, it handles the various raw/non-raw permutations of the literal and format string as well as some other literals. You'd have to change the part where it checks the macro names before it's called + do a lint rename for both `print_literal` and `write_literal` to the new one

**Comment by nyurik** [comment]
@Alexendoo after some more thinking, it seems separating it into a new lint would create more problems than solve:
* multiple lints operating on the same expression tend to cause issues, esp in the `--fix` mode
* a format expression with numeric indexes would be ignored by both lints unless used together: `format!("{0} {1}", var, "str")` will be ignored by both the `uninlined_format_args` and the string inlining lint, but when used together it would suggest the `format!("{var} str")`
* the original lint name `uninlined_format_args` fits perfectly for the string inlining as well, so wouldn't cause any confusion

So in short, we would not really gain anything besides problems from multiple lints...  Not worth it IMO :)

**Comment by Alexendoo** [comment]
I don't see what problem that would cause here, overlapping suggestions would cause rustfix ui tests to fail (`cargo fix` can handle them though) but the suggestions here wouldn't have overlapping spans

The main thing is so that you can set the lint levels independently, e.g. `print_literal` and `write_literal` are `style` but `uninlined_format_args` is `pedantic`. It wouldn't be unusual to want to disable one independently if they were in the same group 

It would also make the lint documentation more confusing to have to explain the two different kinds of inlining in one lint, and how `allow-mixed-uninlined-format-args` would only apply to one of those kinds

`format!("{0} {1}", var, "str")` doesn't seem likely enough to occur to make up for that

**Comment by nyurik** [comment]
@Alexendoo you do raise important config point.  I think we can actually achieve both goals if we track the type of expansion needed -- e.g. if both are enabled, and both are needed (the `"literal", var` case above), we can show just one suggestion (under either of the names, or even add a note saying that both were involved).  If only one is enabled, it won't even consider the other in the expansion.

**Comment by Alexendoo** [comment]
To me that sounds too complicated/novel considering the problem is relatively minor

**Comment by llogiq** [comment]
Agreed, let's not complicate the solution. As long as both lints don't step on each other's toes by issuing incompatible suggestions, having two lints that don't care about each other is fine.

**Comment by xFrednet** [comment]
Hey @nyurik, this is a ping from triage. Do you still plan to continue this PR? If you need help, you're always welcome to ask questions here or on Zulip :)

**Comment by xFrednet** [comment]
Hey this is triage, I'm closing this due to inactivity. If you want to continue this implementation, you're welcome to create a new PR. Thank you for the time, you already put into this!

Interested parties are welcome to pick this implementation up as well :)

@rustbot label +S-inactive-closed -S-waiting-on-author -S-waiting-on-review
